src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -31,34 +96,198 @@ class CoroutineStacksPanel(project: Project) : JBPanelWithEmptyText() {
+                        location,
+                        1,
+                        mutableMapOf(),
+                        "${coroutineData.descriptor.name}${coroutineData.descriptor.id} ${coroutineData.descriptor.state} \n"
+                    )
+                    currentNode.children[location] = node
+                    currentNode = node
+                }
+            }
+        }
+
+        val locationMatrix: MutableList<MutableList<Location?>> = mutableListOf()
+        val stack = Stack<Pair<Node, Int>>()
+        val parentStack = Stack<Node>() // Stack to track parent nodes
+        val headerContent: MutableList<String> = mutableListOf()
+        val hoverContentForHeaders: MutableList<String?> = mutableListOf()
+
+        stack.push(Pair(rootValue, 0))
+        val locationsFollowedBySeparators = mutableListOf<Location?>()
+        var locationMatrixPointer = -1
+
+        while (stack.isNotEmpty()) {
+            val (currentNode, level) = stack.pop()
+            val parent = if (parentStack.isNotEmpty()) parentStack.pop() else null
+
+            if (parent != null) {
+                if (parent.num != currentNode.num) {
+                    headerContent.add("${currentNode.num} Coroutines ACTIVE")
+                    hoverContentForHeaders.add(currentNode.coroutinesActive)
+                    locationMatrix.add(mutableListOf(currentNode.stackFrameLocation))
+                    locationMatrixPointer += 1
+                } else {
+                    locationMatrix[locationMatrixPointer].add(currentNode.stackFrameLocation)
+                }
+            }
+
+            currentNode.children.values.reversed().forEachIndexed { index, child ->
+                if (index != currentNode.children.values.size - 1) {
+                    locationsFollowedBySeparators.add(child.stackFrameLocation)
+                }
+                stack.push(child to (level + 1))
+                parentStack.push(currentNode)
+            }
+        }
+
+        val separator : MutableList<Location?> = mutableListOf(null)
+        val locationMatrixWithSeparators: MutableList<MutableList<Location?>> = mutableListOf()
+        for (locationData in locationMatrix) {
+            if (locationsFollowedBySeparators.contains(locationData[0])) {
+                locationMatrixWithSeparators.add(separator)
+            }
+            locationMatrixWithSeparators.add(locationData)
+        }
+
+        val separatorCount = locationMatrix.size - locationsFollowedBySeparators.size
+        repeat(separatorCount) {
+            locationMatrixWithSeparators.add(separator)
+        }
+
+        val componentData: MutableList<Component> = mutableListOf()
+        var headerIndex = 0
+        locationMatrixWithSeparators.forEachIndexed { i, separatorOrLocation ->
+            if (separatorOrLocation == separator) {
+                componentData.add(Separator())
+            } else {
+                val vertex = JBList<String>()
```

Comment:
Let's create a separate class to handle coroutine stacks UI, say `class CoroutineTrace : JBList<String>`, and handle  
all customisation there